src/Storages/MergeTree/MergeTreeBaseSelectProcessor.cpp

```diff
@@ -628,12 +628,12 @@ Block MergeTreeBaseSelectProcessor::transformHeader(
         else
         {
             WhichDataType which(removeNullable(recursiveRemoveLowCardinality(prewhere_column.type)));
-            if (which.isInt() || which.isUInt())
+            if (which.isNativeInt() || which.isNativeUInt())
```

Comment:
I've checked the test w/o your change, I didn't observe any crash. It was similar error but from other place (stack):
```
2022.10.24 11:52:42.714708 [ 684705 ] {20808803-ee17-4a62-b627-1e68f4398230} <Error> executeQuery: Code: 59. DB::Exception: Illegal type Int128 of column for filter. Must be UInt8 or Nullable(UInt8) or Const variants of them. (ILLEGAL_TYPE_OF_COLUMN_FOR_FILTER) (version 22.10.1.1) (from 127.0.0.1:47244) (in query: SELECT a FROM prewhere_int128 WHERE a;), Stack trace (when copying this message, always include the lines below):

0. ./reldbg/./contrib/libcxx/include/exception:133: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x16dd4c8c in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
1. ./reldbg/./src/Common/Exception.cpp:69: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xeed817a in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
2. ./reldbg/./src/Columns/FilterDescription.cpp:0: DB::FilterDescription::FilterDescription(DB::IColumn const&) @ 0x141a92a3 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
3. ./reldbg/./contrib/libcxx/include/__memory/shared_ptr.h:816: DB::FilterTransform::transformHeader(DB::Block, DB::ActionsDAG const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) @ 0x149ffa0c in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
4. ./reldbg/./contrib/libcxx/include/__memory/shared_ptr.h:816: DB::FilterStep::FilterStep(DB::DataStream const&, std::__1::shared_ptr<DB::ActionsDAG> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool) @ 0x14b25e52 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
5. ./reldbg/./contrib/libcxx/include/__memory/unique_ptr.h:0: DB::InterpreterSelectQuery::executeWhere(DB::QueryPlan&, std::__1::shared_ptr<DB::ActionsDAG> const&, bool) @ 0x13c2b4f5 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
6. ./reldbg/./src/Interpreters/InterpreterSelectQuery.cpp:1519: DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::__1::optional<DB::Pipe>) @ 0x13c23b3b in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
7. ./reldbg/./contrib/libcxx/include/optional:237: DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&) @ 0x13c2240d in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
8. ./reldbg/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:0: DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&) @ 0x13c5fc4b in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
9. ./reldbg/./contrib/libcxx/include/__memory/shared_ptr.h:604: DB::InterpreterSelectWithUnionQuery::execute() @ 0x13c60673 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
10. ./reldbg/./src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x13ebedb7 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
11. ./reldbg/./src/Interpreters/executeQuery.cpp:1106: DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x13ebc20c in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
12. ./reldbg/./src/Server/TCPHandler.cpp:0: DB::TCPHandler::runImpl() @ 0x148c5a64 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
13. ./reldbg/./src/Server/TCPHandler.cpp:1904: DB::TCPHandler::run() @ 0x148d3a59 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
14. ./reldbg/./contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x16dc37e7 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
15. ./reldbg/./contrib/libcxx/include/__memory/unique_ptr.h:54: Poco::Net::TCPServerDispatcher::run() @ 0x16dc3ccd in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
16. ./reldbg/./contrib/poco/Foundation/src/ThreadPool.cpp:213: Poco::PooledThread::run() @ 0x16e362e7 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
17. ./reldbg/./contrib/poco/Foundation/include/Poco/SharedPtr.h:156: Poco::ThreadImpl::runnableEntry(void*) @ 0x16e33f63 in /home/ubuntu/git/3ClickHouse/reldbg/programs/clickhouse
18. ? @ 0x7fdd13121609 in ?
19. __clone @ 0x7fdd13046133 in ?
```

Reply:
It's actually triggered with PREWHERE and it's not a crash but a LOGICAL_ERROR so it crashes on Debug build only.
It can be forced with PREWHERE clause, I'll change the description also.